scripts/weekly-summary

```diff
@@ -0,0 +1,374 @@
+
+    # Find all the test's kstest.log files and append them to kstest.log
+    with open(kstest_log, "w") as ksf:
+        for log in glob(os.path.join(tdir.name, "*", "kstest.log")):
+            with open(log) as f:
+                data = f.read(1024**2)
+                while data:
+                    ksf.write(data)
+                    data = f.read(1024**2)
+
+
+def pass_fail(tests):
+    """
+    pass_fail returns stats on the tests
+
+    It returns a tuple of total tests run, number passed, number failed
+    """
+    fail = 0
+    not_fail = 0
+    for t in tests:
+        if t["success"]:
+            not_fail += 1
+        else:
+            fail += 1
+    return len(tests), not_fail, fail
+
+
+def get_failed(tests):
+    """
+    get_failed returns a dict of failed tests
+
+    Tests may fail multiple times, so they are sorted by their start time
+    and the dict key is the name of the test.
+    """
+    failed = {}
+    for t in tests:
+        if not t["success"]:
+            if t["name"] not in failed:
+                failed[t["name"]] = [t]
+            else:
+                failed[t["name"]].append(t)
+
+    # Sort each failed test list by the start time
+    for t in failed:
+        failed[t] = sorted(failed[t], key=lambda x: x.get("start_time", 0))
+
+    return failed
+
+
+def summary(args, logs):
+    """
+    summary generates a summary of all the tests run in the selected date range
+
+    It returns a string with the summary text
+    """
+    all_data = {}
+    for log in logs:
+        with open(log) as f:
+            data = json.load(f)
+
+            # Group them by group, assume each file is from one group
+            if "group" in data[0]:
+                if data[0]["group"] not in all_data:
+                    all_data[data[0]["group"]] = data
+                else:
+                    all_data[data[0]["group"]].extend(data)
+
+    if args.debug:
+        print(json.dumps(all_data))
+
+    buf = io.StringIO()
+
+    start = args.start.strftime("%m/%d/%Y %H:%M:%S")
+    end = args.end.strftime("%m/%d/%Y %H:%M:%S")
+    print(f"Test Summary Report: {start} -> {end}\n", file=buf)
+
+    # Summary of tests per group
+    for g in sorted(all_data.keys()):
+        total, good, bad = pass_fail(all_data[g])
+        print(f"{g}: Ran {total} tests. {good} passed, {bad} failed.", file=buf)
+    print("\n", file=buf)
+
+    # Test failures per day/group
+    days = {}
+    for g in sorted(all_data.keys()):
+        for t in all_data[g]:
+            if t["success"]:
+                continue
+            day = datetime.fromtimestamp(t.get("start_time", 0)).strftime("%Y%m%d")
+            if day not in days:
+                days[day] = {g: [t["name"]]}
+            else:
+                if g not in days[day]:
+                    days[day][g] = [t["name"]]
+                else:
+                    days[day][g].append(t["name"])
+
+    for day in sorted(days.keys()):
+        if day == "19691231":
```

Comment:
What is this constant?

Reply:
In that case please name it. Also I wonder isn't it already defined in the python as a constant?

Reply:
It's the missing date case from this:
`datetime.fromtimestamp(t.get("start_time", 0)).strftime("%Y%m%d")`

It's not a number, it's a string representation of what fromtimestamp returns when passed a 0. Basically a way to keep any entries without a timestamp collected together. I'll add a note to the fromtimestamp() line to make this clearer.